### PR TITLE
pkcs11-tool: add GOSTR3410 keypair generation

### PR DESCRIPTION
**Pull request by konstantinpersidskiy**
Hello

Recently I found out that pkcs11-tool cannot generate keypairs of mechanism GOSTR3410 (although this mechanism is shown as supported in `pkcs11-tool -M`). This mechanism is a standard mechanism specified by laws in some countries, so I think it is important to make it work  properly in OpenSC.

This commit finishes implementing GOSTR3410 keypair generation feature; now pairs can be generated just like they are with other mechanisms.

Example of keypair generation using native OpenSC pkcs15 framework:
```
$ pkcs11-tool -m GOSTR3410-KEY-PAIR-GEN -k --key-type GOSTR3410:C -d 1234 -l -p '12345678'
Using slot 0 with a present token (0x0)
Key pair generated:
Private Key Object; GOSTR3410 
  PARAMS OID: 06072a850302022303
  label:      Private Key
  ID:         1234
  Usage:      sign
Public Key Object; GOSTR3410 
  PARAMS OID: 06072a850302022303
  VALUE:      044084a5de6a1bfd68ff6b106f6921d4da163244b15ea27dd3a87968d40959a6
              e4170df7674bc7ba098080d7ac2cfe80b29b4ed4369dfedf6e13e42ae32eb7b3
              3233
  label:      Private Key
  ID:         1234
  Usage:      verify
```

Example of keypair generation using another pkcs module:
```
$ pkcs11-tool --module "/home/user/Downloads/librtpkcs11ecp.so" -m GOSTR3410-KEY-PAIR-GEN -k --key-type GOSTR3410:B -d 4321 -l -p '12345678'
Using slot 0 with a present token (0x0)
Key pair generated:
Private Key Object; GOSTR3410 
  PARAMS OID: 06072a850302022302
  label:      
  ID:         4321
  Usage:      sign
Public Key Object; GOSTR3410 
  PARAMS OID: 06072a850302022302
  VALUE:      4c0a3a792a9c03cf227be5b4f94d668a11ce34071b29d9da2585fa1d7cbd1c74
              eb874e468245882c31a66a590cec62699b73b225b69e69ed298ff104be7a8c4f
  label:      
  ID:         4321
  Usage:      verify
```
I believe this commit will be useful. Looking forward for reviews and comments.
Regards, Konstantin Persidskiy

### CODE REVIEW

File: src/libopensc/pkcs15-pubkey.c [hunk 1]
```diff
@@ -1081,6 +1081,8 @@ sc_pkcs15_dup_pubkey(struct sc_context *ctx, struct sc_pkcs15_pubkey *key, struc
 	pubkey->algorithm = key->algorithm;
 
 	if (key->alg_id) {
+		if (key->algorithm == SC_ALGORITHM_GOSTR3410)
+			key->alg_id->params = &(key->u.gostr3410.params);
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-pubkey.c]: Could you please check whether we could improve initializing of the data structures instead?

File: src/pkcs11/framework-pkcs15.c [hunk 2]
```diff
@@ -2741,11 +2741,17 @@ pkcs15_gen_keypair(struct sc_pkcs11_slot *slot, CK_MECHANISM_PTR pMechanism,
...
 	if (keytype == CKK_GOSTR3410)   {
+		unsigned int kp[] = {1, 2, 643, 2, 2, 35, 1, (unsigned int)-1};
+		unsigned int hp[] = {1, 2, 643, 2, 2, 30, 1, (unsigned int)-1};
 		keygen_args.prkey_args.key.algorithm = SC_ALGORITHM_GOSTR3410;
 		pub_args.key.algorithm               = SC_ALGORITHM_GOSTR3410;
 		set_gost_params(&keygen_args.prkey_args.params.gost, &pub_args.params.gost,
 				pPubTpl, ulPubCnt, pPrivTpl, ulPrivCnt);
 		keybits = SC_PKCS15_GOSTR3410_KEYSIZE;
+		memcpy(&keygen_args.prkey_args.key.u.gostr3410.params.key, kp, 8*sizeof(kp[0]));
+		memcpy(&keygen_args.prkey_args.key.u.gostr3410.params.hash, hp, 8*sizeof(kp[0]));
```
Comment (frankmorgner) [inline src/pkcs11/framework-pkcs15.c]: an `int` is not guaranteed to have exactly 8 bytes.

File: src/tools/pkcs11-tool.c [hunk 3]
```diff
@@ -1803,11 +1803,58 @@ static int gen_keypair(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
...
+			
+			if (p_param_set == NULL)
+				util_fatal("Unknown key type %s", type);
+
+			if (!strcmp("A", p_param_set))	
+				memcpy(gost_params, (CK_BYTE[]) {0x06, 0x07, 0x2a, 0x85, 0x03, 0x02, 0x02, 0x23, 0x01}, PARAMS_SIZE);
+			else if (!strcmp("B", p_param_set))	
+				memcpy(gost_params, (CK_BYTE[]) {0x06, 0x07, 0x2a, 0x85, 0x03, 0x02, 0x02, 0x23, 0x02}, PARAMS_SIZE);
+			else if (!strcmp("C", p_param_set))	
+				memcpy(gost_params, (CK_BYTE[]) {0x06, 0x07, 0x2a, 0x85, 0x03, 0x02, 0x02, 0x23, 0x03}, PARAMS_SIZE);
```
Comment (frankmorgner) [inline src/tools/pkcs11-tool.c]: Please don't use magic values here; at least use a good name for them.

File: src/pkcs11/framework-pkcs15.c [hunk 4]
```diff
@@ -2741,11 +2741,17 @@ pkcs15_gen_keypair(struct sc_pkcs11_slot *slot, CK_MECHANISM_PTR pMechanism,
 		goto kpgen_done;
 
 	if (keytype == CKK_GOSTR3410)   {
+		unsigned int kp[] = {1, 2, 643, 2, 2, 35, 1, (unsigned int)-1};
+		unsigned int hp[] = {1, 2, 643, 2, 2, 30, 1, (unsigned int)-1};
```
Comment (frankmorgner) [inline src/pkcs11/framework-pkcs15.c]: please use a better name for your OIDs.

Why are they needed on the PKCS#11 level?

**Comment by frankmorgner** [comment]
Please also check the CI errors

**Comment by konstantinpersidskiy** [comment]
Thanks for the review!
I improved the things you pointed to (and some others). Now the code is more readable and changes are placed in right components.

**Comment by frankmorgner** [comment]
Thanks for the update, please have a look at the CI error.

**Comment by konstantinpersidskiy** [comment]
I already checked new CI errors and they seem to be internal:
- travis reported build to exit with status 0 but didn't finish build job somehow which led to an error
https://travis-ci.org/OpenSC/OpenSC/jobs/226311036
- some of appveyor build jobs failed on downloading files from download.microsoft.com
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/c8j48qfu41ugxvgj
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/e256s3rixg9w0evr
- some others appveyor build jobs couldn't recognize 'nmake' command and failed
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/33sh04db7m5aa593
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/3x2a13931m7e3qyv
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/yikidsla8fah0lgh
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.16.0.1240/job/j6cr0x6aumx83it5

File: src/libopensc/pkcs15.c [hunk 5]
```diff
@@ -856,6 +856,19 @@ sc_pkcs15_card_clear(struct sc_pkcs15_card *p15card)
 }
 
 
+int sc_copy_gost_params(struct sc_pkcs15_gost_parameters *dst, struct sc_pkcs15_gost_parameters *src)
```
Comment (frankmorgner) [inline src/libopensc/pkcs15.c]: Why don't you move this implementation to `src/pkcs15init/pkcs15-lib.c`, which is the only place where this is used?

Comment (konstantinpersidskiy) [inline src/libopensc/pkcs15.c]: Yes, you are right. Moved it.

**Comment by konstantinpersidskiy** [comment]
@frankmorgner , I've made changes to the code you mentioned, can you please check it out?

**Comment by frankmorgner** [comment]
can you be more specific? CI errors were fixed with https://github.com/OpenSC/OpenSC/pull/997/commits/49d876b70e3adbe06a706589cc2447eff792ba95, but I think you didn't look at the length check.

**Comment by konstantinpersidskiy** [comment]
@frankmorgner , I think i don't understand what you mean by "length check".  I thought I've fixed all the code you asked in the series of commits (and the last commit also passes CI checks).
If I collapse these commits in one resulting commit, will that be OK? Is this what you mean? Or otherwise could you please point out what exactly is wrong with my code so I can fix that?

**Comment by frankmorgner** [comment]
sorry, I just realized that the comments were pending and I never finished the review...

**Pull request merged**